resolver/caching_resolver.go

```diff
@@ -172,63 +185,56 @@ func (r *CachingResolver) Resolve(ctx context.Context, request *model.Request) (
 		cacheKey := util.GenerateCacheKey(dns.Type(question.Qtype), domain)
 		logger := logger.WithField("domain", util.Obfuscate(domain))
 
-		val, ttl := r.getFromCache(logger, cacheKey)
+		cacheEntry := r.getFromCache(logger, cacheKey)
 
-		if val != nil {
+		if cacheEntry != nil {
 			logger.Debug("domain is cached")
 
-			val.SetRcode(request.Req, val.Rcode)
-
-			// Adjust TTL
-			setTTLInCachedResponse(val, ttl)
+			cacheEntry.SetRcode(request.Req, cacheEntry.Rcode)
 
-			if val.Rcode == dns.RcodeSuccess {
-				return &model.Response{Res: val, RType: model.ResponseTypeCACHED, Reason: "CACHED"}, nil
+			if cacheEntry.Rcode == dns.RcodeSuccess {
+				return &model.Response{Res: cacheEntry, RType: model.ResponseTypeCACHED, Reason: "CACHED"}, nil
 			}
 
-			return &model.Response{Res: val, RType: model.ResponseTypeCACHED, Reason: "CACHED NEGATIVE"}, nil
+			return &model.Response{Res: cacheEntry, RType: model.ResponseTypeCACHED, Reason: "CACHED NEGATIVE"}, nil
 		}
 
 		logger.WithField("next_resolver", Name(r.next)).Trace("not in cache: go to next resolver")
-		response, err = r.next.Resolve(ctx, request)
 
+		response, err = r.next.Resolve(ctx, request)
 		if err == nil {
-			cacheTTL := r.adjustTTLs(response.Res.Answer)
-			r.putInCache(ctx, cacheKey, response, cacheTTL, true)
+			ttl := r.modifyResponseTTL(response.Res)
+			if ttl > noCacheTTL {
+				cacheCopy := r.putInCache(logger, cacheKey, response)
+				if cacheCopy != nil && r.redisClient != nil {
+					r.redisClient.PublishCache(cacheKey, cacheCopy)
+				}
+			}
```

Comment:
This still adjusts the TTL twice, but if we can get everything down to a single code path, it means `putInCache` can also publish to Redis, which will make this whole part simpler.

Reply:
Maybe wrap it in another function? 🤔

I tried removing the publish Boolean parameter since it seemed misplaced. 
On the other hand also a method to publish without put is needed (prefetching)... 🥴
I'm a little troubled fitting these 3 requirements in one logical block.

Reply:
>  On the other hand also a method to publish without put is needed (prefetching)... 🥴

Not sure if you saw my other comment, but I think we can make this go away without too much trouble.

Reply:
Already commented on it. 👍

Sorry, read and replayed in reverse order. 🫣